internal/broker/broker.go

```diff
@@ -675,42 +676,33 @@ func (b *Broker) loadAuthInfo(ctx context.Context, session *sessionInfo, passwor
 	return authCachedInfo{Token: tok, RawIDToken: refreshedIDToken}, nil
 }
 
-func (b *Broker) fetchUserInfo(ctx context.Context, session *sessionInfo, t *authCachedInfo) (userInfo models.UserInfo, err error) {
+func (b *Broker) fetchUserInfo(ctx context.Context, session *sessionInfo, t *authCachedInfo) (userInfo info.User, err error) {
 	defer decorate.OnError(&err, "could not fetch user info")
 
 	if session.isOffline {
-		return models.UserInfo{}, errors.New("session is in offline mode")
+		return info.User{}, errors.New("session is in offline mode")
 	}
 
 	idToken, err := session.authCfg.provider.Verifier(&b.oidcCfg).Verify(ctx, t.RawIDToken)
 	if err != nil {
-		return models.UserInfo{}, fmt.Errorf("could not verify token: %v", err)
+		return info.User{}, fmt.Errorf("could not verify token: %v", err)
 	}
 
 	userInfo, err = b.providerInfo.GetUserInfo(ctx, t.Token, idToken)
 	if err != nil {
-		return models.UserInfo{}, fmt.Errorf("could not get user info: %v", err)
+		return info.User{}, fmt.Errorf("could not get user info: %v", err)
 	}
 
 	// Some providers are case-insensitive, but we are not. So we need to lowercase the returned username.
 	userInfo.Name = strings.ToLower(userInfo.Name)
 	if userInfo.Name != session.username {
-		return models.UserInfo{}, fmt.Errorf("returned user %q does not match the selected one %q", userInfo.Name, session.username)
+		return info.User{}, fmt.Errorf("returned user %q does not match the selected one %q", userInfo.Name, session.username)
 	}
 
-	return b.fillUserInfoDefaults(userInfo), nil
-}
-
-// fillUserInfoDefaults fills the userinfo struct with default values if they are not set.
-func (b *Broker) fillUserInfoDefaults(u models.UserInfo) models.UserInfo {
-	if u.Shell == "" {
-		u.Shell = "/usr/bin/bash"
-	}
-	if u.Home == "" {
-		u.Home = filepath.Join(b.homeDirPath, u.Name)
+	// This means that home was not provided by the claims, so we need to set it to the broker default.
+	if userInfo.Home == "" {
```

Comment:
I would do that slightly differently:
- in the `info.NewUser()` call, if home claims is empty, set it to `Name`.
- then, here, if userInfo.Home isn’t an absolute path: prepend and join `b.homeDirPath`. That way, we allow:
* providers to provide an absolute path and we respect it
* providers to provide a home directory surrname and we respect it
* still provide a default.

Ofc, we need tests for those :)

Reply:
Well, we do still allow providers to do all of this, no? I'm not sure I like the idea of setting an "invalid" value on the field, as it would basically force us to check the value every time.

Reply:
As said, the 1st one is what will be set by the providers which can override it.

on the second part: well, right now, they set a relative path, we are just taking it and use that "invalid" value :)

This is why at least, here, we would have a contract with the providers:
- send something relative from the broker -> we will prefix it to make it absolute where the config set is and this is a valid use case.
- send something absolute, and we will respect it.
- have a default for complete empty value.

Reply:
answered online.